src/lib/latex/format.rs

```diff
@@ -32,36 +33,154 @@ pub trait Formatter<'source>: Iterator<Item = SpannedToken<'source>> {
+                self.inside_document = true;
+            }
+            Some(&(Token::EnvironmentEnd(_), _)) => {
+                // To count an end envirornment only once
+                if !self.is_indented && self.inside_document {
+                    self.target_indentation_level -= 1;
+                }
+            }
+            _ => {}
+        };
+
+        if !self.is_indented {
+            // Remove current indent
+            if let Some(&(Token::TabsOrSpaces, _)) = self.iter.peek() {
+                self.iter.next();
+                return self.next();
+            }
+
+            self.is_indented = true;
+            let mut indentation_value: String = "".to_owned();
+            for _ in 0..self.target_indentation_level {
+                indentation_value.push_str(&self.indent_chars);
+            }
+
+            // Cannot use .. to define the range because it is a Full Range
+            let custom_indentation: SpannedToken<'source> =
+                (Token::OwnedString(indentation_value), 0..1);
+            Some(custom_indentation)
+        } else {
+            // Post indent matching
+            match self.iter.peek() {
+                Some(&(Token::EnvironmentBegin(_), _)) => {
+                    if self.inside_document {
+                        self.target_indentation_level += 1;
+                    }
+                }
+                Some(&(Token::Newline, _)) => {
+                    self.is_indented = false;
+                }
+                _next => {}
+            };
+            self.iter.next()
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use logos::Logos;
+    use std::io::BufWriter;
+
+    #[test]
+    fn test_document_auto_indent() {
+        let source = r#"
+\documentclass{article}
+  \usepackage{tikz} % Indent in preamble
+
+\begin{test}
+  Nothing should be indented
+\end{test}
+
+\begin{document}
+  Good indentation
+    \begin{tikzpicture}[scale=1.5] % Wrong indentation
+\draw[thick,fill=gray!60] (0,0) rectangle (1,1); % Also wrong
+    \end{tikzpicture}
+It should go back to an indentation level of one
+\end{document}
+
+    % Identation after document
+"#;
+
+        let result = r#"
+\documentclass{article}
+\usepackage{tikz} % Indent in preamble
+
+\begin{test}
+Nothing should be indented
+\end{test}
+
+\begin{document}
+  Good indentation
+  \begin{tikzpicture}[scale=1.5] % Wrong indentation
+    \draw[thick,fill=gray!60] (0,0) rectangle (1,1); % Also wrong
+  \end{tikzpicture}
+  It should go back to an indentation level of one
+\end{document}
+
+% Identation after document
+"#;
+
+        let iter = Token::lexer(source).spanned();
+        let mut buf = BufWriter::new(Vec::new());
+
+        AutoIndentFormatter::new(iter)
+            .write_formatted(source, &mut buf)
+            .unwrap();
+        let bytes = buf.into_inner();
+        let string = String::from_utf8(bytes.unwrap());
+
+        assert_eq!(string.unwrap(), result)
```

Comment:
Why using a BufWriter? Here, I think directly using a `String` makes more sense:
```suggestion
        let mut got = String::with_capacity(result.len());

        AutoIndentFormatter::new(iter)
            .write_formatted(source, &mut got)
            .unwrap();

        assert_eq!(got, result)
```

Reply:
`got` needs to implement the trait `Writable` which is not the case with your current solution.

Reply:
Hum but `String` does implement `Write`: https://doc.rust-lang.org/nightly/std/string/struct.String.html#impl-Write-for-String.

You may have to import it nonetheless.

Reply:
We need the one from `std::io` [https://doc.rust-lang.org/std/io/trait.Write.html](https://doc.rust-lang.org/std/io/trait.Write.html)

Reply:
Ah true, then looks good to me!